TU20Bot/Commands/NameMatch.cs

```diff
@@ -0,0 +1,242 @@
+        /// <param name="firstOrLastMatch">Executed when there is no space in fullname, however either the first or last name match</param>
+        /// <param name="succesfulMatch">Executed when there is a complete match, that is both first name and last name match</param>
+        /// <param name="lastNameMatch">Executed when only the last name matches </param>
+        public static void nameMatchAlg(string fullName, string[,] listNames, IndexMatch firstOrLastMatch, IndexMatchDetailed succesfulMatch, IndexMatchDetailed lastNameMatch)
+        {
+            int spaceIndex = fullName.LastIndexOf(' ');
+
+            for (int i = 0; i < listNames.GetLength(0); i++)
+            {
+                // If a user doesn't have a first or last name in the server
+                if (spaceIndex <= 0)
+                {
+                    if (compare((listNames[i, firstNameIndex] + listNames[i, lastNameIndex]), fullName, false))
+                        firstOrLastMatch(i);
+                }
+
+                /*
+                 * If a user has both first and last names in the server
+                 * If last name matches then proceed to see if the first name matches
+                 * If both of them match then give the user the role passed in the command
+                 * If first name doesn't match then notify in the console that the last name 
+                 *    match was successful.
+                 * If last name doesn't match then notify the user in the console. 
+                 */
+                else
+                {
+                    string firstName = fullName.Substring(0, spaceIndex);
+                    string lastName = fullName.Substring(spaceIndex + 1);
+
+                    // Checking if the last names match
+                    if (compare(listNames[i, lastNameIndex], lastName, true))
+                    {
+                        // Checking if the first names match
+                        if (compare(listNames[i, firstNameIndex], firstName, false))
+                        {
+                            succesfulMatch(i, firstName, lastName);
+                            break;
+                        }
+
+                        lastNameMatch(i, firstName, lastName);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Compares two name strings
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="matchName"></param>
+        /// <param name="fullMatch">If true, then match for equivalency, otherwise match if contains</param>
+        /// <returns></returns>
+        private static bool compare(string name, string matchName, bool fullMatch)
+        {
+            // For last name since last name should be exactly the same
+            if (fullMatch)
+            {
+                return name.Replace(" ", "").Equals(matchName.Replace(" ", "")) || matchName.Replace(" ", "").Equals(name.Replace(" ", ""));
```

Comment:
Generally `a == b` is the same as `b == a`.

```suggestion
                return name.Replace(" ", "") == matchName.Replace(" ", "");
```